server/asset/eth/eth.go

```diff
@@ -0,0 +1,370 @@
+		eth.signalMtx.Lock()
+		for c := range eth.blockChans {
+			select {
+			case c <- &asset.BlockUpdate{
+				Err: err,
+			}:
+			default:
+				eth.log.Errorf("failed to send sending block update on blocking channel")
+				// close(c)
+				// delete(eth.blockChans, c)
+			}
+		}
+		eth.signalMtx.Unlock()
+	}
+
+	sendErrFmt := func(s string, a ...interface{}) {
+		sendErr(fmt.Errorf(s, a...))
+	}
+
+out:
+	for {
+		select {
+		case <-blockPoll.C:
+			tip := eth.blockCache.tip()
+			bestHash, err := eth.node.BestBlockHash(ctx)
+			if err != nil {
+				sendErr(asset.NewConnectionError("error retrieving best block: %w", err))
+				continue
+			}
+			if bestHash == tip.hash {
+				continue
+			}
+
+			block, err := eth.node.Block(ctx, bestHash)
+			if err != nil {
+				sendErrFmt("error retrieving block %x: %w", bestHash, err)
+				continue
+			}
+			// If this doesn't build on the best known block, look for a reorg.
+			prevHash := block.ParentHash()
+			// If it builds on the best block or the cache is empty, it's good to add.
+			if prevHash == tip.hash || tip.height == 0 {
+				eth.log.Debugf("New block %x (%d)", bestHash, block.NumberU64())
+				addBlock(block, false)
+				continue
+			}
+
+			// It is either a reorg, or the previous block is not the cached
+			// best block. Crawl blocks backwards until finding a mainchain
+			// block, flagging blocks from the cache as orphans along the way.
+			//
+			// TODO: Fix this. The exact ethereum behavior here is yet unknown.
+			iHash := tip.hash
+			reorgHeight := uint64(0)
+			for {
+				if iHash == zeroHash {
+					break
+				}
+				iBlock, err := eth.node.Block(ctx, iHash)
+				if err != nil {
+					sendErrFmt("error retrieving block %s: %w", iHash, err)
+					break
+				}
+				// TODO: It is yet unknown how to tell if we are
+				// on the main chain. Blocks do not contain confirmation
+				// information.
+				// if iBlock.Confirmations > -1 {
+				// 	// This is a mainchain block, nothing to do.
+				// 	break
+				// }
+				if iBlock.NumberU64() == 0 {
+					break
+				}
+				reorgHeight = iBlock.NumberU64()
+				iHash = iBlock.ParentHash()
+			}
+
+			var reorg bool
+			if reorgHeight > 0 {
+				reorg = true
+				eth.log.Infof("Tip change from %s (%d) to %s (%d) detected (reorg or just fast blocks).",
+					tip.hash, tip.height, bestHash, block.NumberU64())
+				eth.blockCache.reorg(reorgHeight)
+			}
+
+			// Now add the new block.
+			addBlock(block, reorg)
+
+		case <-ctx.Done():
+			break out
+		}
+	}
+	// Wait for the RPC client to shut down.
+	wg.Wait()
+}
+
+// decodeCoinID decodes the coin ID into a contract address and secret hash.
+func decodeCoinID(coinID []byte) (common.Address, []byte, error) {
+	if len(coinID) != coinIDSize {
+		return common.Address{}, nil, fmt.Errorf("coin ID wrong length. expected %d, got %d",
+			coinIDSize, len(coinID))
+	}
+	secretHash := make([]byte, 32)
+	copy(secretHash, coinID[20:])
+	return common.BytesToAddress(coinID[:20]), secretHash, nil
+}
```

Comment:
I believe a "coin" is best identified by the address of the contract it belongs to and the secret hash it is keyed by. This is tentative and up for debate. See #1001

Reply:
Wouldn't the eth contract address be a constant?

Reply:
yes. If we update the contract, though, I assume it would change.

Reply:
I'm not sure I follow that last part of that. There would not be any need to change a coin ID. Client would initiate the swap with a contract. Server will have a map of contract addresses and reject older contracts if necessary with a penalty. It makes perfect sense that all servers will be using the same contract. I don't think that is a problem.

I believe contracts will need to be updated for security reasons, and if eth consensus rules change.

Reply:
> There would not be any need to change a coin ID

Sorry. Not an individual coin ID. The coin ID encoding protocol. It's hard to change it later.

Reply:
I think I will add a byte for flags. At least we might want unique ids for maker and taker, otherwise they would be the same. If there's somewhere using coin ids as a map key, expecting them to be unique... A one bit flag in the byte field could resolve this, and store a limited amount of arbitrary data in the future. maybe make it two bytes?

Reply:
Added flags anyway. The scheme may be different in the end, as long as we decide on something before allowing mainnet. It may seem like this doesn't allow direct payments for a fee, but, a contract may be also be the best way to handle a fee payment. We could also have a flag that designates this is a transaction hash, then the "secret hash" could instead be the txid in that case.

Reply:
I was wondering about normal transactions too, such as withdraw/deposit and how those would be encoded.  I suspect we will need some kind of dynamic encoding with the flags to switch it to tx hash as you suggested.  I'm just catching up on your eth work though so take that with a grain of salt.

Reply:
Thinking through the coin ID issue: https://github.com/decred/dcrdex/issues/1021